[Workflow] [WIP] Add transition blockers.

d-ph opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #24501
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!--highly recommended for new features-->

- [ ] gather feedback for my changes
- [ ] update docs

Add feature mentioned in #24501.

Please let me know, whether this code is good to merge and I will update the docs for the new feature.

A bit of my commentary: I find supporting `Definitions` with duplicated transition names as a fair maintenance burden for no real benefit. I really can't see any use case why someone might want to have transitions in a workflow, that have duplicated names. In my opinion it violates the principle of least surprise. That is all.

Commit:
[Workflow] #24501 Add transition blockers.

Commit:
[Workflow] #24501 Add transition blockers.

Commit:
[Workflow] #24501 Add transition blockers.

d-ph commented:
The build is failing for unrelated reason (timezone config on the build server?)

d-ph commented on src/Symfony/Component/Workflow/Exception/UndefinedTransitionException.php:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Thrown by Workflow when an undefined transition is applied on a subject.
+ */
+class UndefinedTransitionException extends SubjectTransitionException
```
I created a dedicated exception for this case, so devs can do the following:
```php
try {
  $workflow->apply($subject, $transitionName);
} catch (UndefinedTransitionException $exception) {
  // throw hard
} catch (SubjectTransitionException $exception) {
  // return pretty response with a list of transition blockers.
}
```

d-ph replied:
Ideally, I'd just throw a `LogicException` from all methods, that accept `$transitionName` as a parameter instead of introducing this Exception, but this would break the BC.

d-ph replied:
On second thought, this `try..catch` is annoying, because it would force devs to always catch that exception just to rethrow it. Let me fix that.

d-ph commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -339,4 +358,59 @@ private function announce($subject, Transition $initialTransition, Marking $mark
+     *
+     * @return Transition[]
+     */
+    private function getTransitionsByName(string $transitionName)
```
This is `private`, because I find the fact, that this method returns a collection instead of a `Transition|null` too confusing to make this method `public`,

d-ph commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -339,4 +358,59 @@ private function announce($subject, Transition $initialTransition, Marking $mark
+     * @return Transition[]|TransitionBlockerList All enabled transitions or a blocker list
+     *                                            if no enabled transitions can be found
+     */
+    private function getEnabledTransitionsByNameOrTransitionBlockerList($subject, string $transitionName)
```
Very long name, I agree. This method achieves the same as:
```php
$transitions = $this->getEnabledTransitions();

if (!$transitions) {
  $transitionBlockerList = $this->whyCannot();
}
```
but the benefit is, that this method doesn't run the guards twice in case there are no enabled transitions.

This method wasn't needed until I found out, that unit tests, that test duplicated transition names cases, fail. I.e. this method is an example of the maintenance burden I mentioned in the PR.

lyrixx replied:
I really don't like this method either. Could you get ride of it ?

d-ph replied:
I would like to get rid of it, but I can't because the only other solution is to have the guards be run twice in the pessimistic case scenario (i.e. when there are no enabled transitions). And I'd rather have an ugly but a very dedicated private method than runtime performance degradation.

I guess this is one of those cases, when one looks at other dev's code and thinks "I would've done it better", but when they try to rewrite it to make it better, they realize, there's no better solution.

backbone87 replied:
I would throw an exception that carries the list in case the transition is not enabled. this would be better instead of having multple return types

lyrixx replied:
I agree, But anyway, I'm currently rewrite the Workflow class ATM.
It has become too complex to understand. The very first version was very simple and now it too complex too me.
I will open a new PR today.

d-ph commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -339,4 +358,59 @@ private function announce($subject, Transition $initialTransition, Marking $mark
+        $transitions = array();
+
+        // this is needed to silence static analysis in phpstorm
+        $transitionBlockerList = new TransitionBlockerList();
```
Latest phpstorm can't understand, that at this point there must be at least 1 `$eligibleTransitions`. And I don't like static inspector warnings in source files. If you'd like to remove this line, then I'm fine with that.

Commit:
[Workflow] #24501 Add transition blockers.

Simperfit commented on src/Symfony/Component/Workflow/Exception/SubjectTransitionException.php:
```diff
@@ -22,7 +22,7 @@ class SubjectTransitionException extends LogicException
 
     /**
-     * @param string $message
+     * @param string                $message
```
You can remove the phpdoc

Simperfit commented on src/Symfony/Component/Workflow/Exception/SubjectTransitionException.php:
```diff
@@ -0,0 +1,42 @@
+
+    /**
+     * @param string                $message
+     * @param TransitionBlockerList $transitionBlockerList
```
phpdoc seems useless

nicolas-grekas commented:
This should target 4.1 (4.0/3.4 is in feat freeze), which means you could/should add scalar type hints.

Commit:
[Workflow] #24501 Add transition blockers.

Commit:
[Workflow] #24501 Add transition blockers.

d-ph commented:
@Simperfit 
I removed redundant docblocks.

@nicolas-grekas 
I added scalar and return type hints to the new code. You didn't mention return type hints, but since symfony 4.1. requires php ^7.1.3., I figured that's alright. I can remove the return type hints, if they shouldn't be used. Also, only the new code in `Workflow.php` uses return typehints now, which means that that file has a mixture of methods that define return typehints and that don't define them. I can set the return type hints for the current `Workflow.php::*` methods as well, if you'd like.

lyrixx commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -205,32 +220,35 @@ public function getMarkingStore()
+     *
+     * @return TransitionBlockerList
+     */
     private function doCan($subject, Marking $marking, Transition $transition)
```
Typical exemple of useless PHPDoc ;) 
As this method is private, you can safely add a return type hint

lyrixx commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -205,32 +220,35 @@ public function getMarkingStore()
         foreach ($transition->getFroms() as $place) {
             if (!$marking->has($place)) {
-                return false;
+                return new TransitionBlockerList(array(TransitionBlocker::createNotApplicable($transition->getName())));
```
I have an issue this this new feature.
If the a transition is not enabled because of a missing place in a marking, the end user will only know the **first** reason. It's the same for the guard, they will not be triggered.

Anyway, it's a very bad idea to trigger the guard listener if the marking is not valid (performance issue).

What do you think about that?

d-ph replied:
You're right in saying, that the end user won't know all the blocker reasons, if the the subject isn't in the required state to start with (i.e. the marking is missing a place). I'd say: this is not a bug, but a feature.

I certainly don't need to know the rest of the reasons, if the subject fails to be in the correct state to begin with. And, as you correctly mentioned, it'd be a (massive) performance issue to run the guards in this case, so that's another reason for not doing it.

So what I think is this:
1. Ain't nobody gonna need this (yagni).
2. It's a performance issue.
3. If someone needs this, then he needs to be explained, that this would be a performance issue and therefore it won't be implemented.

lyrixx commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -339,4 +357,59 @@ private function announce($subject, Transition $initialTransition, Marking $mark
+     */
+    private function getTransitionsByName(string $transitionName): array
+    {
+        $transitions = array_filter(
```
you can return directly instead of using a var

d-ph replied:
Yeah, I know. I just have this habit of not immediately returning something, that I might need to xdebug, so I can set a breakpoint on the local variable and see the result of the complex instruction easier.

I changed the code, as you asked.

lyrixx commented on src/Symfony/Component/Workflow/TransitionBlockerList.php:
```diff
@@ -0,0 +1,196 @@
+     *
+     * @param TransitionBlocker[] $blockers The transition blockers to add to the list
+     */
+    public function __construct(array $blockers = array())
```
You never use the array feature here. You always have to write `new TransitionBlockerList([$b]);`
You can change this to:

```php
public function __construct(TransitionBlocker $blocker = null)
{
    if ($blocker) {
        $this->add($blocker);
    }
}
```

d-ph replied:
This constructor expects an array, because `TransitionBlockerList` is an array of `TransitionBlocker`s, but with a dedicated `::findByCode()` method. This method is literally the only reason why I created this class, because otherwise I would just be passing around `TransitionBlocker[]` (i.e. native arrays).

This is also, why there's `ConstraintViolation` and `ConstraintViolationList` in the `Validator` component.

I would be very surprised, if I discovered, that an array-like class can't be constructed with a native array (principle of least surprise).

Also, by accepting an array, I can construct this class with a collection of `TransitionBlocker`s or just a single `TransitionBlocker`. By _not_ accepting an array, I would need to construct this class with a `null` and then manually run the `TransitionBlockerList::add()` in a foreach loop in my own code, if I wanted to construct the class with a collection of `TransitionBlocker`s. This would be extremely annoying (poor DX).

I know, that in the Workflow code I never construct this list with more than 1 item. But taking into account the above, may I leave this constructor with an array argument, please? Having to wrap a single `TransitionBlocker` in an array, in order to construct this class, is really not an issue.

lyrixx reviewed (changes requested):
Hi. 

thanks for this PR. This is a very good start.

I made many comments, I can not review it fully for now. It will be easier after an update.

lyrixx commented:
@d-ph 

> A bit of my commentary: I find supporting Definitions with duplicated transition names as a fair maintenance burden for no real benefit.

I totally agree with you about the maintenance burden. I was against 1+ year ago. But I have changed my mind.

But actually it solves many case with easy. Lets say you have N places: A..N. And you want on each a transition "Go back to A". you can name this transition "Go Back to A". So in you templates / API it's really easier to display "Go back to A" than "Go back to A from B", "Go back to A from C" etc...

And it makes support of StateMachine.

d-ph commented on src/Symfony/Component/Workflow/Exception/SubjectTransitionException.php:
```diff
@@ -0,0 +1,35 @@
+ * Thrown by Workflow when a transition is applied on a subject that is
+ * not possible to be made.
+ */
+class SubjectTransitionException extends LogicException
```
You're right. I went with your name.

d-ph commented on src/Symfony/Component/Workflow/TransitionBlocker.php:
```diff
@@ -0,0 +1,141 @@
+ */
+class TransitionBlocker
+{
+    const REASON_CODE_TRANSITION_NOT_DEFINED = 'com.symfony.www.workflow.transition_blocker.not_defined';
```
Ok.

d-ph commented on src/Symfony/Component/Workflow/TransitionBlocker.php:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * Creates a new transition blocker.
+     *
+     * @param string      $message    The blocker message
```
Ok. I follow the same philosophy in my own code.

d-ph commented on src/Symfony/Component/Workflow/TransitionBlocker.php:
```diff
@@ -0,0 +1,141 @@
+     * @param array       $parameters The parameters that may be useful to pass
+     *                                around with the blocker
+     */
+    public function __construct($message, $code = null, array $parameters = array())
```
How would you construct this class, then? 

The use case for this class is for devs, outside of symfony code, to be able to instantiate TransitionBlockers with their own reasons in their own Guards. This is the whole purpose of this class.

The three reason codes, defined as `const` in this class, are just the reasons that the Workflow itself raises.

I do agree, though, that I'd rather see `$code` not to be nullable. The only reason why it's nullable is to not annoy devs, if they don't care about machine codes. I certainly do, but I don't feel like telling people what's good and what's bad.

When I think of it now, I'd rather just force people to come up with machine codes for their blockers. If they don't care about them, then they can use the `unknown` reason or just fill it with "lorem ipsum".

I'll make the `$code` to be not nullable, but can the constructor remain public, please?

d-ph commented on src/Symfony/Component/Workflow/TransitionBlocker.php:
```diff
@@ -0,0 +1,141 @@
+     */
+    public static function createNotDefined(string $transitionName, string $workflowName)
+    {
+        return new static(
```
Well, what's better:

a) allocating local variables, to avoid using newlines in 1 instruction (runtime overhead)
b) using newlines (no drawbacks)

I'll do what you asked for, though.

d-ph commented on src/Symfony/Component/Workflow/TransitionBlockerList.php:
```diff
@@ -0,0 +1,196 @@
+     *
+     * @throws \OutOfBoundsException if the offset does not exist
+     */
+    public function get($offset): TransitionBlocker
```
There is a use case for just getting the first transition blocker and showing that as an error. I, for example, rarely block transitions for more than 1 reason in my own code, so having a fast and easy way (i.e. an easy `::get(0)` method) is a plus for me, so I don't need to remember, that I can "cast" the iterator to an array and retrieve the first blocker from there. In other words, remembering `::get(0);` is less mental effort than remembering `iterator_to_array($var)[0]`.

You can't assume, that all people will want to output all errors or even have more than 1 error at a time. I, for example, use both approaches interchangeably when I clearly see the benefit of doing one over the other.

I agree, that the "set" part ArrayAccess is useless. I do appreciate, though, that I can use the array-access operator to retrieve the first transition blocker. In such cases and in my code, I just throw from the offsetSet() and offsetUnset(). But I know, that a lot of people disagree with this practice. Can I instead remove the `ArrayAccess` methods, but keep the `::get()` one, please?

d-ph commented on src/Symfony/Component/Workflow/TransitionBlockerList.php:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    /**
+     * Creates iterator for blockers with specific code.
```
I agree, that the doc block is wrong. It's a copy&paste failure from `ConstraintViolationList.php`. I've removed the doc block, because it's obvious what that method does.

It's useful to keep this method as non-internal, because then devs can run conditional logic (i.e. `if` statements) in their code, basing on presence of a specific TransitionBlocker in the TransitionBlockerList, that's carried over from the `Workflow::apply()` method. I, for example, am going to check for the presence of `TransitionBlocker::REASON_TRANSITION_NOT_APPLICABLE` in my code and handle this case slightly differently, than other blockers.

It's also useful to keep this method as a getter and not as a hasser, because the getter form allows devs to see, whether the specific blocker is there and also retrieve this blocker, while the hasser only allows to see, whether the blocker is there. It's useful to be able to retrieve the blocker, because it may contain `$parameters`, that matter to the dev.

This method returns the first blocker, because I assumed, that no one would try to block a transition with the same reason twice. Although possible, I don't know why someone would need to do that. I certainly don't need that.

In summary: can I keep the method as it is, please (except for the wrong doc block of course)?

d-ph commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -92,24 +94,29 @@ public function getMarking($subject)
-            if ($transitionName === $transition->getName() && $this->doCan($subject, $marking, $transition)) {
-                return true;
-            }
+        if ($transitionsOrTransitionBlockerList instanceof TransitionBlockerList) {
```
I actually like my method name better as it's shorter and still clearly says what it does, at least in my opinion. But I changed that, as you asked.

Commit:
[Workflow] #24501 Add transition blockers.

Commit:
[Workflow] #24501 Add transition blockers.

backbone87 commented on src/Symfony/Component/Workflow/TransitionBlocker.php:
```diff
@@ -0,0 +1,102 @@
+    {
+        $message = sprintf('Transition "%s" is not defined in workflow "%s".', $transitionName, $workflowName);
+
+        return new static($message, self::REASON_TRANSITION_NOT_DEFINED);
```
workflow and transition name could be additionally passed as parameters.

backbone87 commented on src/Symfony/Component/Workflow/TransitionBlocker.php:
```diff
@@ -0,0 +1,102 @@
+     */
+    public static function createNotApplicable(string $transitionName): self
+    {
+        $message = sprintf('Transition "%s" cannot be made, because the subject is not in the required place.', $transitionName);
```
transition name could be additionally passed as parameter.

backbone87 commented on src/Symfony/Component/Workflow/TransitionBlocker.php:
```diff
@@ -0,0 +1,102 @@
+     */
+    public static function createUnknownReason(string $transitionName): self
+    {
+        $message = sprintf('Transition "%s" cannot be made, because of unknown reason.', $transitionName);
```
transition name could be additionally passed as parameter.

d-ph replied:
You're right. I guess I didn't eat my own dog food here. I've done what you mentioned in all three places.

Commit:
[Workflow] #24501 Add transition blockers.

Commit:
[Workflow] #24501 Add transition blockers.

Commit:
[Workflow] #24501 Add transition blockers.

Commit:
[Workflow] #24501 Add transition blockers.

lyrixx commented on src/Symfony/Component/Workflow/Event/GuardEvent.php:
```diff
@@ -11,21 +11,52 @@
+    }
 
-    public function isBlocked()
+    public function getTransitionBlockerList(): TransitionBlockerList
```
Changing the order of method make the diff harder to read and to review. Could you keep the same method order ?

Thanks

d-ph replied:
The order of methods didn't change. Github's diff tool just struggles to find the similarity, due to the amount of changes in this file. Screenshot of PhpStorm's diff tool:

![image](https://user-images.githubusercontent.com/4347095/35484635-72a0f7d0-044a-11e8-97a1-b9d0e898caca.png)

Let me know, if I missed something.

lyrixx commented on src/Symfony/Component/Workflow/Exception/BlockedTransitionException.php:
```diff
@@ -0,0 +1,35 @@
+use Symfony\Component\Workflow\TransitionBlockerList;
+
+/**
+ * Thrown by Workflow when a transition is applied on a subject that is
```
> Thrown by the workflow when a transition is not enabled.

(I'm not a native English speaker, but it seems easier to understand)

d-ph replied:
I agree. I just wanted to use the vocabulary set out by the `Workflow` component. 

I'll use the sentence you suggested.

lyrixx commented on src/Symfony/Component/Workflow/TransitionBlocker.php:
```diff
@@ -0,0 +1,112 @@
+     * in a workflow.
+     *
+     * @param string $transitionName
+     * @param string $workflowName
```
this PHP Doc is useless. Please remove it (same for other occurrence)

d-ph replied:
This specific one -- yes, it's useless.

I don't find the remaining two useless, though. 

1. `createNotApplicable()` alone doesn't include the fact, that it's supposed to be created only, when a transition isn't applicable due to the subject's being in wrong status. For any other reason, why transition isn't applicable, there should be a different transition blocker used (i.e. a different blocker code). I can rename the method to `createNotApplicableDueToWrongSubjectPlace()` to make it more clear, if you'd like me to.
1. `createUnknownReason()`, the docblock for this one emphasizes, that this transition blocker is used for preserving backwards compatibility and, by extension, shouldn't be used in new code.

I kept the docblock for `::createNotDefined()` to keep it consistent with other methods' having docblocks in this file. 

If after this explanation you still would like me to remove these docblocks, then I'll remove them.

lyrixx replied:
Sorry, I was talking only about the parameter.
(Note I'm going to finish your PR ;)

lyrixx commented on src/Symfony/Component/Workflow/TransitionBlocker.php:
```diff
@@ -0,0 +1,112 @@
+     */
+    public static function createNotDefined(string $transitionName, string $workflowName): self
+    {
+        $message = sprintf('Transition "%s" is not defined in workflow "%s".', $transitionName, $workflowName);
```
Since PHP 7.0, you can use: `$message = "Transition '$transitionName' is not defined in workflow '$workflowName'."`

It's easier to read IMHO

backbone87 replied:
sprintf is used throughout symfony, as well as double quotes for quoting values in messages

d-ph replied:
Yeah, I used `sprintf()` because it's used everywhere else. If I recall correctly, I used string interpolation at first, but then was told to change it to `sprintf()` by fabbot.io.

lyrixx commented on src/Symfony/Component/Workflow/TransitionBlockerList.php:
```diff
@@ -0,0 +1,107 @@
+/**
+ * A list of transition blockers.
+ */
+class TransitionBlockerList implements \IteratorAggregate, \Countable, \ArrayAccess
```
Do we really need the ArrayAccess ? I don't think so

backbone87 replied:
I agree ArrayAccess should be dropped and maybe even an interface introduced that only grants read access

lyrixx replied:
:-1: for the interface.

d-ph replied:
I'll remove the `ArrayAccess`, but I found that useful to think of the instances of this class as "smarter arrays".

I'll also remove the `::remove()` method, since I didn't find it useful in any use case from the very beginning.

I disagree, that the thing should be completely immutable, though. The common pattern in Transition Guards is to start with an empty `TransitionBlockerList` and then `::add()` blockers as `if` statements evaluate, that the transition isn't possible.

Commit:
[Workflow] Add transition blockers.

backbone87 commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -81,47 +83,57 @@ public function getMarking($subject)
     {
-        $transitions = $this->definition->getTransitions();
-        $marking = $this->getMarking($subject);
+        return 0 === count($this->buildTransitionBlockerList($subject, $transitionName));
```
Since you use this logic in quite a few places, it would be nice to have a TransitionBlockerList::isEmpty

backbone87 commented on src/Symfony/Component/Workflow/WorkflowInterface.php:
```diff
@@ -42,6 +44,16 @@ public function getMarking($subject);
+     *
+     * @return TransitionBlockerList Empty if the transition is possible
+     */
+    public function buildTransitionBlockerList($subject, string $transitionName): TransitionBlockerList;
```
isnt this a BC break? AFAIK you are only allowed to add new interface methods in major versions

lyrixx replied:
No, because it's only in SF 4.1 which has not been released.

lyrixx commented:
Fixed in favor of #26076
